lib/include/pl/core/ast/ast_node_type_decl.hpp

```diff
@@ -76,15 +76,46 @@ namespace pl::core::ast {
         [[nodiscard]] std::vector<std::shared_ptr<ptrn::Pattern>> createPatterns(Evaluator *evaluator) const override {
             evaluator->updateRuntime(this);
 
-            evaluator->pushTemplateParameters();
-            ON_SCOPE_EXIT { evaluator->popTemplateParameters(); };
+            std::vector<std::unique_ptr<ASTNodeLiteral>> templateParamLiterals(this->m_templateParameters.size());
 
-            for (const auto &templateParameter : this->m_templateParameters) {
+            // Get template parameter values before pushing a new template parameter scope so that variables from the parent scope are used before being overwritten.
+            for (size_t i = 0; i < this->m_templateParameters.size(); i++) {
+                auto &templateParameter = this->m_templateParameters[i];
                 if (auto lvalue = dynamic_cast<ASTNodeLValueAssignment *>(templateParameter.get())) {
+                    if (!lvalue->getRValue())
+                        err::E0003.throwError(fmt::format("There is no value for template parameter variable {}. This is a bug.", lvalue->getLValueName()), {}, this);
```

Comment:
I'd rename this error to something like `No value set for non-type template parameter {}. This is a bug.`